// universal-ws/dist/access.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const session_1 = require("./session");
const events_1 = require("events");
var session_2 = require("./session");
exports.StatusCode = session_2.StatusCode;
class UniversalWebSocket extends events_1.EventEmitter {
    constructor(host, connectionOptions, ...parameters) {
        super();
        this.state = session_1.State.closed;
        this.session = new session_1.Session(host, connectionOptions, ...parameters);
        this.heartbeatMode = this.session.heartbeatMode;
        this.heartbeatInterval = this.session.heatbeatInterval;
        this.responseTimeout = this.session.responseTimeout;
        this.session.on('connected', () => {
            this.emit('connected');
        });
        this.session.on('disconnected', (code, reason) => {
            this.emit('disconnected', code, reason);
        });
        this.session.on('message', (message, data) => {
            this.emit(message, data);
        });
        this.session.on('request', (message, data, callback) => {
            this.emit(message, data, callback);
        });
        this.session.on('state', (state) => {
            this.state = state;
            this.emit('state', this.state);
        });
        this.session.on('error', (error) => {
            this.emit('error', error);
        });
    }
    negotiate(settings) {
        return this.session.negotiate(settings);
    }
    send(message, data) {
        this.session.send(message, data);
    }
    sendWithAck(message, data) {
        return this.session.sendWithAck(message, data);
    }
    request(message, data) {
        return this.session.request(message, data);
    }
    close(code = session_1.StatusCode.Normal_Closure, reason) {
        return this.session.close(code, reason);
    }
    open() {
        return this.session.open();
    }
}
exports.UniversalWebSocket = UniversalWebSocket;
