// universal-ws/dist/transport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const bs58 = require('bs58');
const DELIMITER = '$';
var Type;
(function (Type) {
    Type[Type["Browser"] = 0] = "Browser";
    Type[Type["Node"] = 1] = "Node";
})(Type = exports.Type || (exports.Type = {}));
var ReadyState;
(function (ReadyState) {
    ReadyState[ReadyState["Connecting"] = 0] = "Connecting";
    ReadyState[ReadyState["Open"] = 1] = "Open";
    ReadyState[ReadyState["Closing"] = 2] = "Closing";
    ReadyState[ReadyState["Closed"] = 3] = "Closed";
})(ReadyState = exports.ReadyState || (exports.ReadyState = {}));
var StatusCode;
(function (StatusCode) {
    StatusCode[StatusCode["Normal_Closure"] = 1000] = "Normal_Closure";
    StatusCode[StatusCode["Going_Away"] = 1001] = "Going_Away";
    StatusCode[StatusCode["Protocol_Error"] = 1002] = "Protocol_Error";
    StatusCode[StatusCode["Unexpected_Data"] = 1003] = "Unexpected_Data";
    StatusCode[StatusCode["Reserved_For_Future"] = 1004] = "Reserved_For_Future";
    StatusCode[StatusCode["No_Status_Code_Present"] = 1005] = "No_Status_Code_Present";
    StatusCode[StatusCode["Invalid_Data"] = 1006] = "Invalid_Data";
    StatusCode[StatusCode["Message_Error"] = 1007] = "Message_Error";
    StatusCode[StatusCode["Message_Too_Large"] = 1008] = "Message_Too_Large";
    StatusCode[StatusCode["Unexpected_Error"] = 1011] = "Unexpected_Error";
})(StatusCode = exports.StatusCode || (exports.StatusCode = {}));
class UniversalWs {
    constructor(host, options) {
        // https://developer.mozilla.org/en-US/docs/Web/API/WebSocket
        if (typeof WebSocket !== 'undefined') {
            this.ws = options.parameters ? new WebSocket(host, [this.encodeParameters(options.parameters)]) : new WebSocket(host);
            return;
        }
        try {
            const ws = require('ws');
            if (ws) {
                this.ws = options.parameters ? new ws(host, [this.encodeParameters(options.parameters)], options.perMessageDeflateOptions) : new ws(host, options.perMessageDeflateOptions);
                return;
            }
        }
        catch (_a) {
            throw new Error('Cannot construct WebSocket! Your environment may not support web sockets. See: https://caniuse.com/#feat=websockets');
        }
    }
    encodeParameters(parameters) {
        return bs58.encode(Buffer.from(parameters.join(DELIMITER), 'utf8'));
    }
    on(eventName, callback) {
        if (!this.ws)
            return;
        if (isBrowser(this.ws)) {
            switch (eventName) {
                case 'open':
                    this.ws.addEventListener('open', (event) => callback(event));
                    break;
                case 'message':
                    this.ws.addEventListener('message', (event) => {
                        callback(event.data);
                    });
                    break;
                case 'close':
                    this.ws.addEventListener('close', (event) => callback({ code: event.code, reason: event.reason }));
                    break;
                case 'error':
                    this.ws.addEventListener('error', (event) => callback(event));
                    break;
                default:
                    throw (`Invalid event name ${eventName}`);
            }
        }
        else {
            switch (eventName) {
                case 'open':
                    this.ws.on('open', callback);
                    break;
                case 'message':
                    this.ws.on('message', (data) => callback(data));
                    break;
                case 'close':
                    this.ws.on('close', (code, reason) => callback({ code, reason }));
                    break;
                case 'error':
                    this.ws.on('error', (error) => callback(error));
                    break;
                default:
                    throw (`Invalid event name ${eventName}`);
            }
        }
    }
    send(message) {
        if (!this.ws)
            return;
        if (isBrowser(this.ws)) {
            this.ws.send(message);
        }
        else {
            this.ws.send(message);
        }
    }
    close(code = 1000, reason = '') {
        if (!this.ws)
            return;
        if (isBrowser(this.ws)) {
            if (this.ws.readyState === this.ws.OPEN)
                this.ws.close(code, reason);
        }
        else {
            this.ws.close(code, reason);
        }
    }
    getReadyState() {
        if (!this.ws)
            return;
        if (isBrowser(this.ws)) {
            return ReadyState[this.ws.readyState];
        }
        else {
            return ReadyState[this.ws.readyState];
        }
    }
}
exports.UniversalWs = UniversalWs;
function isBrowser(ws) {
    return typeof WebSocket !== 'undefined';
}
